include/xtensor/xrepeat.hpp

```diff
@@ -0,0 +1,442 @@
+                {
+                    ++inner_steps;
+                    m_index += m_repeats;
+                }
+                return inner_steps;
+            }
+
+            void reset()
+            {
+                m_index = 0;
+            }
+
+            void reset_back()
+            {
+                m_index = m_repeats - 1;
+            }
+
+        private:
+            const std::ptrdiff_t m_repeats;
+            std::ptrdiff_t m_index;
+        };
+
+    private:
+        S m_substepper;
+        std::vector<repeated_index> m_indicies;
+    };
+
+    template <class CT>
+    class xrepeat :
+        public xiterable<xrepeat<CT>>,
+        public xaccessible<xrepeat<CT>>
+    {
+    public:
+        using xexpression_type = std::decay_t<CT>;
+        using value_type = typename xexpression_type::value_type;
+        using shape_type = typename xexpression_type::shape_type;
+
+        using container_type = xcontainer_inner_types<xrepeat<CT>>;
+        using reference = typename container_type::reference;
+        using const_reference = typename container_type::const_reference;
+        using size_type = typename container_type::size_type;
+        using temporary_type = typename container_type::temporary_type;
+
+        static constexpr layout_type static_layout = xexpression_type::static_layout;
+        using bool_load_type = typename xexpression_type::bool_load_type;
+        using pointer = typename xexpression_type::pointer;
+        using const_pointer = typename xexpression_type::const_pointer;
+        using difference_type = typename xexpression_type::difference_type;
+
+        using iterable_type = xiterable<xrepeat<CT>>;
+        using stepper = typename iterable_type::stepper;
+        using const_stepper = typename iterable_type::stepper;
+
+        template<class CTA>
+        explicit xrepeat(CTA&& e,
+            std::vector<std::ptrdiff_t>&& repeats,
+            std::vector<std::ptrdiff_t>&& axes);
+
+        template <class... Args>
+        reference operator()(Args... args);
+
+        template <class... Args>
+        const_reference operator()(Args... args) const;
+
+        const shape_type& shape() const noexcept;
+
+        template <class It>
+        reference element(It first, It last);
+
+        template <class It>
+        const_reference element(It first, It last) const;
+
+        stepper stepper_begin(const shape_type& s) const;
+
+        stepper stepper_end(const shape_type& s, const layout_type l) const;
+
+    private:
+        CT m_e;
+        std::vector<std::ptrdiff_t> m_repeat_lookup;
+        shape_type m_shape;
+
+        reference access();
+
+        template <class Arg, class... Args>
+        reference access(Arg arg, Args... args);
+
+        const_reference access() const;
+
+        template <class Arg, class... Args>
+        const_reference access(Arg arg, Args... args) const;
+
+        template<typename std::decay_t<CT>::size_type I>
+        struct access_impl
```

Comment:
Why not using `access_impl` methods instead of an inner class?

Reply:
I had to split up the template argument `I` from the template arguments `Arg` and `Args...`.
When I not split up those arguments, I have to specify the arguments in the call:

```
s.step(I, arg);
access_impl<I+1, Arg, Args...>(args...);
```

Unfortunately, `Arg` will not be the type of the first argument of `args...`.
Therefore, this snippet will not work.

May you can show me another approach to solve this problem.

Reply:
I think you can specify the `size_type` argument only and let the compiler deduce the other ones:

```cpp
s.step(I, arg);
access_impl<I+1>(args...);
```